aten/src/ATen/native/cuda/Dropout.cu

```diff
@@ -16,8 +20,84 @@ namespace {
 
 // philox generates 128 bits of randomness at a time. Kernel uses this explicitly by putting suitably transformed result into float4
 // for all members of float4 to be consumed UNROLL has to be 4. Don't change!
+// Note: VEC <= 4 (and in most real-world cases will be 4), so same logic applies.
 const int UNROLL = 4;
 
+template <
+          typename scalar_t,
+          typename accscalar_t,
+          typename IndexType,
+          int ADims,
+          int VEC>
+#if __CUDA_ARCH__ >= 350
+C10_LAUNCH_BOUNDS_2(256, 8)
+#elif defined (__HIP_PLATFORM_HCC__)
+C10_LAUNCH_BOUNDS_2(256, 4)
+#endif
+__global__ void
+fused_dropout_kernel_vec(at::cuda::detail::TensorInfo<scalar_t, IndexType> a,
+                            at::cuda::detail::TensorInfo<scalar_t, IndexType> b,
+                            at::cuda::detail::TensorInfo<uint8_t, IndexType> c,
+                            IndexType totalElements, accscalar_t p, std::pair<uint64_t, uint64_t> seeds
+                           ) {
+  typedef typename std::aligned_storage<VEC*sizeof(scalar_t), VEC*alignof(scalar_t)>::type LoadT;
+  typedef typename std::aligned_storage<VEC*sizeof(uint8_t), VEC*alignof(uint8_t)>::type MaskLoadT;
+
+  accscalar_t pinv = accscalar_t(1)/p;
+  IndexType idx = blockIdx.x * blockDim.x + threadIdx.x;
+  curandStatePhilox4_32_10_t state;
+  curand_init(
+      seeds.first,
+      idx,
+      seeds.second,
+      &state);
+
+  // Note: Vectorized loads means we'll stride each thread by an additional VEC factor, as we'll load VEC elements at a time
+  for (IndexType linearIndex = idx * VEC;
+      linearIndex < totalElements;
+      linearIndex += gridDim.x * blockDim.x * VEC) {
+    // local storage
+    scalar_t src[VEC];
+    // We'll use this to actually cause vectorized loads later
+    LoadT *value = reinterpret_cast<LoadT*>(&src);
+
+    //curand_uniform_double was pure evil anyway, not doing what it promises, and there's nothing for halfs, so generate float for everything
+    // Note: need a new set of random values per 4 elements -- we'll handle VEC elements in this thread, so need ceil(VEC / 4)
+    // sets of rand.
+    float4 rand = curand_uniform4(&state);
+
+    rand.x = rand.x < p;
+    rand.y = rand.y < p;
+    rand.z = rand.z < p;
+    rand.w = rand.w < p;
+
+    // Note: We explicitly check for is_contiguous() before launching the vectorized kernel
+    // -- the IndexToOffset call _could_ be replaced with just linearIndex, but this is
+    // hopefully low enough overhead to not matter
+    const IndexType aOffset =
+      cuda::detail::IndexToOffset<scalar_t, IndexType, ADims>::get(linearIndex, a);
+    // Single vectorized load
+    *value = *reinterpret_cast<LoadT*>(&a.data[aOffset]);
+
+    scalar_t r[VEC];
+    uint8_t mask[VEC];
```

Comment:
bool tensors are now supported, so you could use it to store mask, but up to you.